constantine/math/arithmetic/finite_fields_precomp_square_root.nim

```diff
@@ -148,51 +147,44 @@ func invSqrtEqDyadic*(z: var Fp): bool =
 
   ## looking at the dlogs, powers[i] is essentially the wanted exponent, left-shifted by i*_sqrtBlockSize and taken mod 1<<32
   ## dlogHighDyadicRootNeg essentially (up to sign) reads off the _sqrtBlockSize many most significant bits. (returned as low-order bits)
   ## 
   ## first iteration may be slightly special if BlockSize does not divide 32
-  negExponent = sqrtAlg_NegDlogInSmallDyadicSubgroup(powers[Fp.C.tonelliShanks(sqrtParam_Blocks) - 1])
-  negExponent = negExponent shr Fp.C.tonelliShanks(sqrtParam_FirstBlockUnusedBits)
+  negExponent = sqrtAlg_NegDlogInSmallDyadicSubgroup(powers[Fp.C.sqrtDlog(Blocks) - 1])
+  negExponent = negExponent shr Fp.C.sqrtDlog(FirstBlockUnusedBits)
 
   # if the exponent we just got is odd, there is no square root, no point in determining the other bits
-  if (negExponent and 1) == 1:
-    return false
+  # if (negExponent and 1) == 1:
+  #   return false
+
+  # result = SecretBool((negExponent and 1) != 1)
 
-  for i in 1..<Fp.C.tonelliShanks(sqrtParam_Blocks):
-    temp2 = powers[Fp.C.tonelliShanks(sqrtParam_Blocks) - 1 - i]
+  for i in 1..<Fp.C.sqrtDlog(Blocks):
+    temp2 = powers[Fp.C.sqrtDlog(Blocks) - 1 - i]
     for j in 0..<i:
-      sqrtAlg_GetPrecomputedRootOfUnity(temp, int( (negExponent shr (j*Fp.C.tonelliShanks(sqrtParam_BlockSize))) and Fp.C.tonelliShanks(sqrtParam_BitMask) ), uint(j + Fp.C.tonelliShanks(sqrtParam_Blocks) - 1 - i))
+      sqrtAlg_GetPrecomputedRootOfUnity(temp, int( (negExponent shr (j*Fp.C.sqrtDlog(BlockSize))) and Fp.C.sqrtDlog(BitMask) ), uint(j + Fp.C.sqrtDlog(Blocks) - 1 - i))
       temp2.prod(temp2, temp)
     
     var newBits = sqrtAlg_NegDlogInSmallDyadicSubgroup(temp2)
-    negExponent = negExponent or (newBits shl ((i*Fp.C.tonelliShanks(sqrtParam_BlockSize)) - Fp.C.tonelliShanks(sqrtParam_FirstBlockUnusedBits)))
+    negExponent = negExponent or (newBits shl ((i*Fp.C.sqrtDlog(BlockSize)) - Fp.C.sqrtDlog(FirstBlockUnusedBits)))
 
   negExponent = negExponent shr 1
   z.setOne()
 
-  for i in 0..<Fp.C.tonelliShanks(sqrtParam_Blocks):
-    sqrtAlg_GetPrecomputedRootOfUnity(temp, int((negExponent shr (i*Fp.C.tonelliShanks(sqrtParam_BlockSize))) and Fp.C.tonelliShanks(sqrtParam_BitMask)), uint(i))
+  for i in 0..<Fp.C.sqrtDlog(Blocks):
+    sqrtAlg_GetPrecomputedRootOfUnity(temp, int((negExponent shr (i*Fp.C.sqrtDlog(BlockSize))) and Fp.C.sqrtDlog(BitMask)), uint(i))
     z.prod(z, temp)
 
-  return true
-
-func sqrtPrecomp*(dst: var Fp, x: Fp): SecretBool {.inline.} =
+func inv_sqrt_precomp*(dst: var Fp, x: Fp) {.inline.} =
   dst.setZero()
-  if x.isZero().bool():
-    return SecretBool(true)
-  
-  var xCopy: Fp = x
   var candidate, rootOfUnity: Fp
-  sqrtAlg_ComputerRelevantPowers(xCopy, candidate, rootOfUnity)
-  if not invSqrtEqDyadic(rootOfUnity):
-    return SecretBool(false)
-
+  sqrtAlg_ComputeRelevantPowers(x, candidate, rootOfUnity)
+  invSqrtEqDyadic(rootOfUnity)
   dst.prod(candidate, rootOfUnity)
-  return SecretBool(true)
-
+  dst.inv()
```

Comment:
Assuming the input `x`

The result of `invSqrtEqDyadic(rootOfUnity)`

should be `x⁻¹ᐟ²` according to https://github.com/GottfriedHerold/Bandersnatch/blob/7c0464f7dfae8f1139c9f812d1543de631f5262b/bandersnatch/fieldElements/field_element_square_root.go#L273-L277
```go
// invSqrtEqDyadic asserts that z is a 2^32 root of unity and tries to set z := 1/sqrt(z).
//
// If z is actually a 2^32th *primitive* root of unity, the square root does not exist and we return false without modifying z.
// Otherwise, z is changed to 1/sqrt(z) and we return true
func (z *feType_SquareRoot) invSqrtEqDyadic() (ok bool) {
```

then the last two lines do

```
dst <- candidate * x⁻¹ᐟ²
dst <- dst⁻¹ (and this is equal to x⁻¹ᐟ² because it matches with invsqrt)
```

Ergo, `candidate = x` and `rootOfUnity = x⁻¹ᐟ²` before the last 2 lines, so they are unnecessary.

This can be raised as a refactoring issue and then cleaned up afterwards.

Reply:
I have seen this while implementing, that's why I haven't commented this particular function.
The problem is when I try to reverse this from the back, then it doesn't work which implies `x != candidate`, which actually should be

But again if `sqrtX = candidate * x⁻¹ᐟ²`, then candidate should be equal to `x`. I have been confused initially with this, but left this to proceed with the implementation

Reply:
I see, then we can leave it as is and scheduled that as a next refactoring.

Reply:
A similar trick to Tonelli-Shanks as described here can probably be applied:

https://github.com/mratsim/constantine/blob/5894a8da757bb935b681c73d1fdc92a7838f94ed/sage/square_root_bls12_377.sage#L222-L248